### ISSUE TITLE
3d dimension but real matrix 

### ISSUE THREAD

**Issue by MADONOKOUKI** [HazyResearch/learning-circuits#11]
I use Butterfly matrix when input size is like (128,256,64).
But due to the dimension, I can't use thie tensor....
How should I fix the code ?

Also As this website mentioned, (https://dawn.cs.stanford.edu/2019/06/13/butterfly/#compression)
I want to sample the pemutation matrix from butterfly matrices in order to take smoothness loss, but can't find the code

**Comment by tridao** [comment]
You can reshape the input to combine the first 2 dimensions before passing through the butterfly layer:
```
butterfly = Butterfly(64, 64)
b, m, n = x.shape
output = butterfly(x.reshape(-1, n)).reshape(x.shape)
```
Re: permutation code: we're in the process of cleaning up that code for release.

**Comment by MADONOKOUKI** [comment]
thanks! It very cool.....
Now I'm trying to train the permutated image whose number of blocks are 64 with Butterfly matrix + ResNet.

But if I only use one Butterfly(64, 64) layer, is it mean that the Butterfly(64, 64) will be permutation matrix ? (LIke Fig7.'s 4 red buttefly matrices?)
https://dawn.cs.stanford.edu/2019/06/13/butterfly/#compression

By the way, if you aren't mind, I want to use code of making permutation matrix which is made by Butterfly matrices

**Comment by tridao** [comment]
I'm not sure I understand what "number of blocks" mean.
Figure 7 contains multiple sparse matrices, each is called a *butterfly factor*. The product of those log_n butterfly factor is called a *butterfly matrix*. So `Butterfly(64, 64)` is a product of `log 64 = 6` butterfly factors.
To capture a permutation matrix of size 64, you'll need a butterfly and the transpose of another butterfly:
```
torch.nn.Sequential(
    Butterfly(64, 64, bias=False, tied_weight=False, increasing_stride=False),
    Butterfly(64, 64, bias=False, tied_weight=False, increasing_stride=True)
)
```
For the code above, `increasing_stride=False` means it's a transpose of a butterfly, and `increasing_stride=True` means it's a butterfly matrix. The flag `tied_weight` should generally be set to `False` for increased expressivity. The default value of `tied_weight=True` is just for compatibility with the experiments in our ICML19 paper (https://arxiv.org/abs/1903.05895).

PS: I'll try to find the permutation code today or tomorrow. Hopefully that'll give you a better idea of how to implement permutation sampling with butterfly.

**Comment by MADONOKOUKI** [comment]
I tried to say "number of blocks" is like a number of puzzle pieces!
I understand the structure of the Butterfly matrix! thank you and sorry for taking your time...
Thank you for sharing the code, although I didn't fully understand the mathematical equations on the ICML19 paper (https://arxiv.org/abs/1903.05895), try to read the paper by confirming the code and understand these functions!!

Thank you very much!! because this paper may be applied to my project, It is very helpful m(_ _)mm(_ _)mm(_ _)mm(_ _)m

**Comment by albertfgu** [comment]
To use a (soft) permutation matrix, you want a Butterfly matrix where each 2x2 block in the factors are constrained to be doubly stochastic (bullet 1 in the blog post). This [version](https://github.com/HazyResearch/learning-circuits/blob/cnn/butterfly/butterfly.py) of the class accepts argument `param='ds'`, `param='logit'`, or `param='ortho2'`, all of which represent doubly stochastic butterfly matrices. (The latter two options map the underlying parameter through the sigmoid or cos^2 function, respectively.)
[Here's](https://github.com/HazyResearch/learning-circuits/blob/cnn/cnn/models/presnet.py) an example of experimental code which implements the ResNet with a butterfly layer in front.

**Comment by MADONOKOUKI** [comment]
Thank you for telling how to use this code!!
If I use PResNet for assuming that input is puzzle images and recover original state, is is adequate to set method='butterfly' in class  TensorPermutation(Permutation) and set param='ds', param='logit', or param='ortho2' in class ButterflyPermutation(Permutation)?

**Comment by albertfgu** [comment]
The latter is the most flexible way to create the latent doubly stochastic matrix.
